### DISCUSSION TITLE
run dag-processor as a job, only once

### DISCUSSION THREAD

**Discussion by antonio-antuan** [apache/airflow#36436] [Ideas] [answered]
In our installation we don't need to run dag-processor periodically, it's enough to run it once right after deployment.
Currently dag-processor does not support in charts something like "-n 1". Actually it is supported but when process is finished - pod is restarted. I assume that the right way to do it is to run it as a job.

Is it possible to do it without any changes in airflow's helm chart?

**Comment by potiuk** [comment, marked as answer]
No. You need to run dag file processor continuously. This is required by airflow. One of the reasons is that in certain circumstances dag file processor also runs callbacks for tasks that have been forcefully killed - they are not only used to serialize DAGs - but also to preform this housekeeping while DAG to execute is parsed to memory as Python object. 


Currently airflow does not support the configuration where dag file processor (or scheduler) periodic parsing is disabled.  We do not plan to change it currently but maybe in the future we will change it. For now it's a basic assumption that dag file processor runs continuously. Disabling it will cause strange problems

**Reply by antonio-antuan** [reply]
I see, thanks. 
What if I increase period of dags scanning? let's say, once an hour. Does it affect those callbacks in a way that they will be executed only once an hour when this "scanning" process run?

**Reply by potiuk** [reply]
Yes

**Reply by tokoko** [reply]
@potiuk sorry for digging up an old discussion. What's you current stance on this? I think this will make a lot of sense once multi-team feature lands. We have been running our stitched-together version of multi-team internally (a central webserver/scheduler/db, plus each team with their own dag-processors and workers). One problem that we came across recently is that some of those "teams" (they are effectively more like projects, to be completely honest) have so small of a workload (a couple of dags) it can hardly justify having a considerable long-running deployments just for them. The solution we're currently experimenting with is to run dag-processor as a k8s job (`-n 1`) on each deploy and launch tasks in k8s executor rather than celery, avoiding the need for running deployments for either a dag processor or a worker.

Of course this presupposes that dag code doesn't depend on anything that can be changed w/o a commit in the repo and you don't care about some of the callbacks that are still ran by dag-processor. So far we only came across an issue when scheduler deactivates our dags unless we set `dag_stale_not_seen_duration` config to a huge value.

**Reply by potiuk** [reply]
Airlfow 3 brought "dag bundle" feature and multi-team (latest incarnation of it) builds on top of it - each team will be able to have their own dag bundle. And the design of the dag bundle foresees (though it's coming in the future versions of airlfow - likely the same that multi-team) that for each bundle you will be able to configure the frequency of updates - including "on change". After this is implemented - yes you will be able to run a dedicated dag processor "per team" and have that dag processor run only once, because dag processor will not be doing those continuous parsing/reparsing if you configure your dag bundle to do so.

That is coming.
